<?xml version="1.0"?>
<!-- Note: example launch file. Do not modify! -->
<launch>
  <arg name="gui" default="true" />

  <!-- The following nodes are excluded from the group since they use move_group
  from MoveIt!, which remaps the 'move_group' namespace to the group's namespace -->
  <node pkg="mcr_manipulation_pose_selector" type="reachability_pose_selector_node"
          name="pregrasp_pipeline_pose_selector" output="screen">

    <!-- the node cycle rate (in hz) -->
    <param name="loop_rate" type="double" value="10.0"/>

    <!-- MoveIt! node name -->
    <param name="move_group" type="str" value="move_group"/>

    <!-- Arm name to be used for giving commands-->
    <param name="arm" type="str" value="arm_1"/>

    <rosparam param="linear_offset">
      - 0.00
      - 0.00
      - 0.00
    </rosparam>

    <remap from="~goal_pose_array"
           to="/mir_manipulation/pregrasp_planner_pipeline/pose_generator/poses_list"/>
    <remap from="~joint_configuration"
           to="/mir_manipulation/pregrasp_planner_pipeline/configuration_out"/>
    <remap from="~selected_pose"
           to="/mir_manipulation/pregrasp_planner_pipeline/selected_pose"/>
  </node>

  <group ns="mir_manipulation/pregrasp_planner_pipeline">
    <node pkg="mcr_manipulation_measurers" type="pose_transformer_node"
              name="pose_transformer" output="screen">

      <!-- the cycle rate of the node (in Hz) -->
      <param name="loop_rate" type="double" value="10.0"/>

      <param name="wait_for_transform" type="double" value="0.1"/>

      <!-- Reference frame to transform target pose -->
      <param name="target_frame" type="str" value="base_link"/>

      <!-- Flag to use node in trigger mode -->
      <param name="trigger" type="boolean" value="false"/>

      <!-- Pose in (geometry_msgs/PoseStamped) -->
      <remap from="~pose_in" to="/mir_manipulation/pregrasp_planner_pipeline/target_pose" />
    </node>

    <node pkg="mir_pregrasp_planning" type="simple_pregrasp_planner_node"
            name="pregrasp_planner" output="screen">

      <!-- the node cycle rate (in hz) -->
      <param name="loop_rate" type="double" value="10.0"/>

      <!-- the tolerance to decide whether an object should be re-oriented,
      based on its height (in meters) -->
      <param name="height_tolerance" type="double" value="0.08"/>

      <!-- Closest distance the gripper should be to the object (in meters) -->
      <param name="min_distance_to_object" type="double" value="0.0"/>
      <!-- Furthest distance the gripper should be to the object (in meters) -->
      <param name="max_distance_to_object" type="double" value="0.0"/>

      <!-- Azimuth (in degrees) -->
      <param name="min_azimuth" type="double" value="-3.0"/>
      <param name="max_azimuth" type="double" value="3.0"/>

      <!-- Zenith (in degrees) -->
      <param name="min_zenith" type="double" value="-3.0"/>
      <param name="max_zenith" type="double" value="3.0"/>

      <!-- Wrist roll (in degrees) -->
      <param name="min_roll" type="double" value="90.0"/>
      <param name="max_roll" type="double" value="90.0"/>

      <!-- Angular tolerance to check if an object is standing up (in degrees) -->
      <param name="angular_tolerance" type="double" value="10.0"/>

      <remap from="~pose_in" to="pose_transformer/transformed_pose"/>
      <remap from="~grasp_type" to="/mir_manipulation/pregrasp_planner_pipeline/grasp_type"/>
    </node>

    <node pkg="mcr_pose_generation" type="pose_generator_node"
            name="pose_generator" output="screen">

      <rosparam command="load"
                file="$(find mir_pregrasp_planning)/ros/config/gripper_configuration_matrix.yaml" />

      <!-- the cycle rate of the node (in hz) -->
      <param name="loop_rate"  type="double" value="10.0"/>

      <!-- the sampling step for linear variables (in meters) -->
      <param name="linear_step" type="double" value="0.01"/>

      <!-- the sampling step for angular variables (in degrees) -->
      <param name="angular_step" type="double" value="0.5"/>

      <!-- the configuration matrix of the gripper to be used -->
      <param name="gripper_config_matrix" type="str" value="youbot_dynamixel_gripper"/>

      <remap from="~target_pose"
             to="/mir_manipulation/pregrasp_planner_pipeline/pregrasp_planner/pose_out"/>
      <remap from="~sampling_parameters"
             to="/mir_manipulation/pregrasp_planner_pipeline/pregrasp_planner/sampling_parameters"/>
    </node>

    <node pkg="mir_pregrasp_planning" type="pregrasp_planner_pipeline_node"
            name="coordinator" output="screen">

      <!-- the node cycle rate (in hz) -->
      <param name="loop_rate" type="double" value="30.0"/>

      <remap from="~event_in" to="/mir_manipulation/pregrasp_planner_pipeline/event_in"/>
      <remap from="~event_out" to="/mir_manipulation/pregrasp_planner_pipeline/event_out"/>

      <!-- To trigger components -->
      <remap from="~start_pose_transformer" to="pose_transformer/event_in"/>
      <remap from="~start_pregrasp_planner"
             to="/mir_manipulation/pregrasp_planner_pipeline/pregrasp_planner/event_in"/>
      <remap from="~start_pose_generator"
             to="/mir_manipulation/pregrasp_planner_pipeline/pose_generator/event_in"/>
      <remap from="~start_pose_selector"
             to="/pregrasp_pipeline_pose_selector/event_in"/>

      <!-- To get feedback from components -->
      <remap from="~pose_transformer_status" to="pose_transformer/event_out"/>
      <remap from="~pregrasp_planner_status"
             to="/mir_manipulation/pregrasp_planner_pipeline/pregrasp_planner/event_out"/>
      <remap from="~pose_generator_status"
             to="/mir_manipulation/pregrasp_planner_pipeline/pose_generator/event_out"/>
      <remap from="~pose_selector_status"
             to="/pregrasp_pipeline_pose_selector/event_out"/>
    </node>
  </group>

  <!-- GUI Helper(s) -->
  <group if="$(arg gui)">
    <node pkg="mir_pregrasp_planning" type="pose_mock_up_gui_node"
            name="pose_mock_up_gui" output="screen">

      <!-- the cycle rate of the node (in hz) -->
      <param name="loop_rate" type="double" value="10.0"/>
      <param name="reference_frame" type="str" value="base_link"/>

      <remap from="~mock_up_pose" to="/mir_manipulation/pregrasp_planner_pipeline/target_pose"/>
    </node>
  </group>
</launch>
